drivers/serial/uart_emul.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2023 Fabian Blatz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_uart_emul
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/drivers/serial/uart_emul.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_emul, CONFIG_UART_LOG_LEVEL);
+
+struct uart_emul_config {
+};
+
+struct uart_emul_data {
+	struct uart_config cfg;
+
+	struct ring_buf rx_rb;
+	uint8_t rx_buf[CONFIG_UART_EMUL_BUFFER_SIZE];
+	struct k_spinlock rx_lock;
+
+	struct ring_buf tx_rb;
+	uint8_t tx_buf[CONFIG_UART_EMUL_BUFFER_SIZE];
+	struct k_spinlock tx_lock;
+};
+
+int uart_emul_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	struct uart_emul_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	if (ring_buf_is_empty(&data->rx_rb)) {
+		LOG_DBG("Rx buffer is empty");
+		return -1;
```

Comment:
Use error codes from `errno.h`

Reply:
I hardcoded it since this is an expected return value of the uart interface according to the docs. But I changed it to `-ENOBUFS`.